### PR TITLE
masq-profiles replication and data hyperdb replication 

### PR DESCRIPTION
**Pull request by leventdem**
See #2

### CODE REVIEW

File: test/masq.test.js [hunk 1]
```diff
@@ -1,36 +1,85 @@
+const MasqCore = require('./MasqCore')
+// const rai = require('random-access-idb')
```
Comment (G-Ray) [inline test/masq.test.js]: remove unused imports

File: test/masq.test.js [hunk 2]
```diff
@@ -1,36 +1,85 @@
...
 
 test('initialize', async () => {
   masq = new Masq()
   await masq.init()
+  masqCore = new MasqCore()
+  await masqCore.init()
+  await masqCore.initProfiles()
+  await masqCore.setProfile('bob', profile1)
+  const profile = await masqCore.getProfile('bob')
+  expect(profile).toEqual(profile1)
```
Comment (G-Ray) [inline test/masq.test.js]: You are testing masqCore, which is a mock, this is a non-sense, be sure to focus only on the masq-lib, which is the lib the developers will integrate in their apps

File: test/masq.test.js [hunk 3]
```diff
@@ -100,29 +149,77 @@ test('should be kicked if challenge does not match', async (done) => {
...
+  })
+
+  masqCore.on('changeDataHello', async msg => {
+    await wait()
+    // check if item is replicated in masq
+    const res = await masqCore.getItem(appName, '/hello')
+    expect(res).toEqual({ data: 'world' })
+    const sw = masqCore.getDataSw(appName)
+    const hub = masqCore.getDataHub(appName)
+    sw.close()
```
Comment (G-Ray) [inline test/masq.test.js]: The test shouldn't be in charge of handling swarms and hubs, the Masq mock should

File: test/masq.test.js [hunk 4]
```diff
@@ -1,36 +1,85 @@
...
+  image: 'image1'
+}
 beforeAll(() => {
   server = signalserver()
   server.listen(8080, () => console.log('server running'))
 })
 
-afterAll(() => {
+afterAll(async () => {
+  await wait()
```
Comment (G-Ray) [inline test/masq.test.js]: why do we need to wait here ?

File: test/masq.test.js (hunk 2 above)
Comment (G-Ray) [reply test/masq.test.js]: If you need to setup the Masq mock with some data, do it in a `beforeAll/beforeEach`

File: test/masq.test.js [hunk 5]
```diff
@@ -100,29 +149,77 @@ test('should be kicked if challenge does not match', async (done) => {
...
+    channel: channel,
+    challenge: challenge,
+    appName: appName
+  })
+
+  masqCore.on('changeDataHello', async msg => {
+    await wait()
+    // check if item is replicated in masq
+    const res = await masqCore.getItem(appName, '/hello')
+    expect(res).toEqual({ data: 'world' })
```
Comment (G-Ray) [inline test/masq.test.js]: Again, you are testing the mock instead of the masq lib

File: src/index.js [hunk 6]
```diff
@@ -95,48 +111,195 @@ class Masq {
...
+            const db = hyperdb(rai('masq-profiles'), Buffer.from(json.key, 'hex'), { valueEncoding: 'json' })
+            await dbReady(db)
+
+            // Store
+            this.dbs.profiles = db
+
+            peer.send(JSON.stringify({
+              msg: 'replicationProfilesStarted'
+            }))
+            // db replication
```
Comment (G-Ray) [inline src/index.js]: If everything is ok, we should close the hub and sw here right ?

File: src/index.js [hunk 7]
```diff
@@ -95,48 +111,195 @@ class Masq {
...
+            peer.send(JSON.stringify({
+              msg: 'requestWriteAccess',
+              key: db.local.key.toString('hex')
+            }))
+          }
+          break
+        case 'ready':
+          // Masq must send ready after the authorization
+          this._startDataReplication()
+          sw.close()
```
Comment (G-Ray) [inline src/index.js]: missing `hub.close()` too

File: src/index.js [hunk 8]
```diff
@@ -95,48 +111,195 @@ class Masq {
...
-    const json = JSON.parse(data)
-    switch (json.msg) {
-      case 'challenge':
-        if (json.challenge !== this.challenge) {
-          // This peer may be malicious, close the connection
-          this.sw.close()
-          this.hub.close()
-        }
-        break
+  async _startDataReplication () {
```
Comment (G-Ray) [inline src/index.js]: _startReplication and _startDataReplication seems to share 99% of code, can we refactor to only one function ?

File: test/masq.test.js [hunk 9]
```diff
@@ -100,8 +106,127 @@ test('should be kicked if challenge does not match', async (done) => {
...
+  sw.on('disconnect', (peer, id) => {
+    sw.close()
+    hub.close()
+  })
+
+  const _handleData = async (data, peer) => {
+    const json = JSON.parse(data)
+    switch (json.msg) {
+      case 'requestWriteAccess':
+        expect(json.key).toHaveLength(64)
```
Comment (G-Ray) [inline test/masq.test.js]: good

File: test/masq.test.js [hunk 10]
```diff
@@ -126,3 +251,24 @@ test('del should reject when there is no profile selected', async () => {
...
   }
 })
+
+test('put/get should put and get an item', async () => {
+  expect.assertions(1)
+  const key = '/hello'
+  const value = { data: 'world' }
+  masq.setProfile(profile.id)
+  dbTest = hyperdb(rai(profile.id), { valueEncoding: 'json' })
+  // Only for test purpose, we overwrite the data hyperdb
```
Comment (G-Ray) [inline test/masq.test.js]: perfect :+1:

File: test/masq.test.js [hunk 11]
```diff
@@ -126,3 +251,24 @@ test('del should reject when there is no profile selected', async () => {
...
+  const res = await masq.get('/hello')
+  expect(res).toEqual(value)
+})
+
+test('del should del an item', async () => {
+  expect.assertions(1)
+  const key = '/hello'
+  await masq.del(key)
+  const res = await masq.get('/hello')
+  expect(res).toBeUndefined()
```
Comment (G-Ray) [inline test/masq.test.js]: :+1:

File: src/index.js (hunk 6 above)
Comment (leventdem) [reply src/index.js]: yes, added ws.close()

File: src/index.js (hunk 7 above)
Comment (leventdem) [reply src/index.js]: not useful because hub.close will be called automatically

File: src/index.js (hunk 8 above)
Comment (leventdem) [reply src/index.js]: yes, done

File: test/masq.test.js (hunk 9 above)
Comment (leventdem) [reply test/masq.test.js]: thanks.

File: src/index.js [hunk 12]
```diff
@@ -95,48 +111,164 @@ class Masq {
...
+      peer.on('data', data => _handleData(data, peer))
+    })
+
+    sw.on('close', () => {
+      hub.close()
+    })
+
+    sw.on('disconnect', (peer, id) => {
+      sw.close()
+      hub.close()
```
Comment (G-Ray) [inline src/index.js]: Should not be needed as the sw.on('close') event will close the hub too

File: src/index.js [hunk 13]
```diff
@@ -95,48 +111,164 @@ class Masq {
...
+      pump(peer, stream, peer)
+    })
+
+    sw.on('close', () => {
+      hub.close()
+    })
+
+    sw.on('disconnect', (peer, id) => {
+      sw.close()
+      hub.close()
```
Comment (G-Ray) [inline src/index.js]: Same, it will be closed in the close event

File: test/masq.test.js [hunk 14]
```diff
@@ -100,8 +106,127 @@ test('should be kicked if challenge does not match', async (done) => {
...
+  const sw = swarm(hub, { wrtc })
+  const key = 'c4b60362325d27ad3c04db158fa68fe6fde00387467708ab3a0be79c811b3825'
+
+  sw.on('peer', (peer, id) => {
+    peer.on('data', data => {
+      const json = JSON.parse(data)
+
+      expect(json.msg).toBe('replicationProfilesStarted')
+      sw.close()
+      hub.close()
```
Comment (G-Ray) [inline test/masq.test.js]: same

File: test/masq.test.js [hunk 15]
```diff
@@ -100,8 +106,127 @@ test('should be kicked if challenge does not match', async (done) => {
...
+    }))
+  })
+
+  sw.on('close', () => {
+    hub.close()
+  })
+
+  sw.on('disconnect', (peer, id) => {
+    sw.close()
+    hub.close()
```
Comment (G-Ray) [inline test/masq.test.js]: Same

File: test/masq.test.js [hunk 16]
```diff
@@ -100,8 +106,127 @@ test('should be kicked if challenge does not match', async (done) => {
...
+  })
+
+  const _handleData = async (data, peer) => {
+    const json = JSON.parse(data)
+    switch (json.msg) {
+      case 'requestWriteAccess':
+        expect(json.key).toHaveLength(64)
+        // authorize local key & start replication
+        sw.close()
+        hub.close()
```
Comment (G-Ray) [inline test/masq.test.js]: again :laughing:

File: src/index.js [hunk 17]
```diff
@@ -95,48 +111,164 @@ class Masq {
...
     }
+    const sw = this.sws[name]
+
+    sw.on('peer', async (peer, id) => {
+      const stream = db.replicate({ live: true })
+      pump(peer, stream, peer)
+    })
+
+    sw.on('close', () => {
+      hub.close()
```
Comment (G-Ray) [inline src/index.js]: you also need to delete this.sws[name] and this.hubs[name]

File: src/index.js (hunk 13 above)
Comment (leventdem) [reply src/index.js]: done

File: test/masq.test.js (hunk 15 above)
Comment (leventdem) [reply test/masq.test.js]: done

File: test/masq.test.js (hunk 16 above)
Comment (leventdem) [reply test/masq.test.js]: done

File: test/masq.test.js (hunk 14 above)
Comment (leventdem) [reply test/masq.test.js]: done

File: src/index.js (hunk 12 above)
Comment (leventdem) [reply src/index.js]: done

File: src/index.js (hunk 17 above)
Comment (leventdem) [reply src/index.js]: Added after hub.close()

**Pull request merged**